README.md

```diff
@@ -1,2 +1,228 @@
+])
+
+module.exports = config
+```
+
+```js
+// index.js
+//
+// The following env vars can be set:
+//
+// LOGLEVEL - defaults to info
+// PORT - default to 3000
+//
+
+'use strict'
+
+const http = require('http')
+const config = require('./config.js')
+
+// This validates that we have the necessary env vars.
+config.validateEnvVars()
+
+http.listen(config.get('port'), () => {
+  log.info('listen', config.get('port'))
+})
+```
+
+Under the hood, `Config` is a [`<Map>`][], so use it like one.
+
+This package also provides a way to automatically generate documentation
+for the environment variables for a service.
+If the `doc` directory does not exist, it can be created using the following:
+
+```
+$ mkdir -p doc
+```
+
+Then, add a `generate` script to the
+service's `package.json` that looks similar to this:
+
+```sh
+config-doc config.js > doc/env.md
+```
+
+You should also add a link to this document in the `README.md` of the service.
+
+## API
+
+### `new Config(input)`
+
+* `input` [`<Array>`][] Array of objects that represent a single rule.
+
+Each `input` item should be a `Definition`. The following static methods
+return a `Definition` that can be used.
+
+#### `Config.string(name)`
+
+* `name` [`<String>`][] Name of the config rule. The environment variable
+read for this rule will be `name` uppercased with `-` replaced with `_`.
+
+This defines a configuration definition for an environment variable that is
+a [`<String>`][].
+
+Returns: `Definition`
+
+#### `Config.number(name)`
+
+* `name` [`<String>`][] Name of the config rule. The environment variable
+read for this rule will be `name` uppercased with `-` replaced with `_`.
+
+This defines a configuration defintion for an environment variable that is
+a [`<Number>`][].
+
+Returns: `Definition`
+
+#### `Config.boolean(name)`
+
+* `name` [`<String>`][] Name of the config rule. The environment variable
+read for this rule will be `name` uppercased with `-` replaced with `_`.
+
+This defines a configuration definition for an environment variable that is
+a [`<Boolean>`][].
+
+Returns: `Definition`
+
+#### `Config.regex(name)`
+
+* `name` [`<String>`][] Name of the config rule. The environment variable
+read for this rule will be `name` uppercased with `-` replaced with `_`.
+
+This defines a configuration definition for an environment variable that
+matches a [`<RegExp>`][].
+
+Returns: `Definition`
+
+#### `Config.enum(name)`
+
+* `name` [`<String>`][] Name of the config rule. The environment variable
+read for this rule will be `name` uppercased with `-` replaced with `_`.
+
+A single value will be read from the environment variable as described,
+but it must exist in the allowed list of [`.values()`](#definitionvaluesarr).
+Every `enum` type needs to implement a `.values()` array.
+
+#### `Config#toJSON()`
+
+Returns a JSON representation of the config object.
+This will only work for Primitive values, objects, and arrays.
+
+#### `Config#validateEnvVars()`
```

Comment:
What does the # signify here in `Config#validateEnvVars()`- also am seeing this elsewhere.

Reply:
It's a standard way (in many languages) to denote class methods or members.  I think for javascript it has roots in `jsdoc`

> package.class#member is any valid program element name that is referenced -- a package, class, interface, constructor, method or field name -- except that the character ahead of the member name should be a hash character (#)